File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for (auto &v: board){
            unordered_set<int> nums;
            for (auto c: v){
                if (c != '.'){
                    if (nums.find(c - '0') != nums.end()){
                        return false;
                    } else{
                        nums.insert(c - '0');
                    }
                }
            }
        }

        for (int i = 0; i < 9; ++i){
            unordered_set<int> colNums;
            for (auto &v: board){
                if (v[i] != '.'){
                    if (colNums.find(v[i] - '0') != colNums.end()){
                        return false;
                    } else{
                        colNums.insert(v[i] - '0');
                    }
                }
            }
        }

        unordered_set<int> boxNums[9];
        for (int i = 0; i < 9; ++i){
            for (int j = 0; j < 9; ++j){
                int boxInd = (i / 3) * 3 + (j / 3);
                if (board[i][j] != '.'){
                    if (boxNums[boxInd].find(board[i][j]) != boxNums[boxInd].end()){
                        return false;
                    } else{
                        boxNums[boxInd].insert(board[i][j]);
                    }
                }
            }
        }



        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def checkBox(self, board, row, col) -> bool:
        nums = set()
        for i in range(row, row+3):
            for j in range(col, col+3):
                if board[i][j] != '.':
                    if board[i][j] in nums:
                        return False
                    nums.add(board[i][j])
        return True

    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for row in range(9):
            nums = set()
            for col in range(9):
                if board[row][col] != '.':
                    if board[row][col] in nums:
                        return False
                    nums.add(board[row][col])
        
        for col in range(9):
            nums = set()
            for row in range(9):
                if board[row][col] != '.':
                    if board[row][col] in nums:
                        return False
                    nums.add(board[row][col])
            
        for row in range(3):
            for col in range(3):
                if self.checkBox(board, row * 3, col * 3) == False:
                    return False

        return True
